File: frontend/src/app/features/self-evaluation/self-evaluation.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import {
  SelfEvaluationRequest,
  SelfEvaluationResponse,
} from '../../shared/models/self-evaluation.model';

/**
 * HTTP access to the /api/self-evaluations endpoints.
 */
@Injectable({ providedIn: 'root' })
export class SelfEvaluationService {
  private readonly http = inject(HttpClient);

  private readonly baseUrl = `${environment.apiUrl}/self-evaluations`;

  /**
   * Submits the answers of an employee. This endpoint is public.
   *
   * @param request answers typed in the form
   * @returns the stored self evaluation, including its risk level
   */
  submit(request: SelfEvaluationRequest): Observable<SelfEvaluationResponse> {
    return this.http.post<SelfEvaluationResponse>(this.baseUrl, request);
  }

  /**
   * Reads a single self evaluation with its answers.
   *
   * @param id identifier of the self evaluation
   * @returns the self evaluation
   */
  findById(id: number): Observable<SelfEvaluationResponse> {
    return this.http.get<SelfEvaluationResponse>(`${this.baseUrl}/${id}`);
  }

  /**
   * Reads the self evaluations submitted by the employees of a company.
   *
   * @param companyId identifier of the company
   * @returns list of self evaluations
   */
  findByCompany(companyId: number): Observable<SelfEvaluationResponse[]> {
    return this.http.get<SelfEvaluationResponse[]>(this.baseUrl, { params: { companyId } });
  }
}

File: Backend/src/main/java/com/mgs/ergomanager/service/SelfEvaluationService.java
package com.mgs.ergomanager.service;

import com.mgs.ergomanager.dto.selfevaluation.SelfEvaluationRequestDTO;
import com.mgs.ergomanager.dto.selfevaluation.SelfEvaluationResponseDTO;
import com.mgs.ergomanager.model.enums.RiskLevel;
import java.util.List;

/**
 * Self evaluations filled in by the employees of the client companies.
 */
public interface SelfEvaluationService {

    /**
     * Stores a self evaluation and calculates its risk level.
     *
     * @param request answers sent by the employee
     * @return the stored self evaluation, including its risk level
     */
    SelfEvaluationResponseDTO submit(SelfEvaluationRequestDTO request);

    /**
     * Returns a single self evaluation with its answers.
     *
     * @param id identifier of the self evaluation
     * @return the self evaluation
     */
    SelfEvaluationResponseDTO findById(Long id);

    /**
     * Returns the self evaluations submitted by the employees of a company.
     *
     * @param companyId identifier of the company
     * @return list of self evaluations
     */
    List<SelfEvaluationResponseDTO> findByCompany(Long companyId);

    /**
     * Calculates the risk level that matches a total score.
     *
     * @param totalScore sum of the weighted answer scores
     * @return risk level of the self evaluation
     */
    RiskLevel calculateRiskLevel(Integer totalScore);
}
